rfcs/immediate-arrays.md

```diff
@@ -0,0 +1,222 @@
+  except for some very rare workflows
+
+- ... but the *perception of cost* is troubling our users and
+  encouraging them to consider sub-par implementation choices
+  (see the 'Alternatives' section below).
+
+I propose to solve this situation by adding specific support in the
+Array module for arrays of immediate values, at a *different type*
+than the usual type `'a array`, using a restricted submodule in
+a style reminiscent of Array.Float (but documented).
+
+
+## Proposed interface
+
+An `Array.MakeImmediate` functor with the following signature:
+
+```ocaml
+MakeImmediate :
+functor
+  (Elt : sig type t [@@immediate] end)
+-> sig
+  type elt = Elt.t
+  type t (** arrays of immediate values *)
+  
+  val make : int -> elt -> t
+  val length : t -> int
+
+  val get : t -> int -> elt
+  val unsafe_get : t -> int -> elt
+  
+  val set : t -> int -> elt -> unit
+  val unsafe_set : t -> int -> elt -> unit  
+end
+```
+
+Proposed implementation: a custom block.
+
+The intention is to provide a minimal interface for a low-level
+module, rather than trying (like Float.Array) of copying the whole
+interface of Array. I would propose to restrict the scope of this
+submodule with this intention, but accept proposals/requests for
+extensions when they concern primitives that can be implemented
+significantly more efficiently in C -- for example `fill`,
+`blit` and `sub`.
+
+#### Design choice: a functor on immediate types
+
+This could be specialized to just an `IntArray` module, but I believe
+that a functor over any immediate type is significantly more
+convenient for end-users. In particular, this allows its usage on
+types such as
+
+```ocaml
+type id = Id of int [@@unboxed]
+```
+
+which improve type-abstraction and safety.
+
+Note: one could consider a design where we accept an `immediate64`
+type as parameter, with an implementation that reverts to a standard
+array type on 32-bit machines. `immediate64` is obscure and inconvient
+to use from the stdlib, so I dropped the idea; we could provide an
+additional `MakeImmediate64` version in the future if we really wanted
+to.
+
+#### Design note: asking for an initial element
+
+We could skip the demand for an initial element and 0-initialize the
+array. I decided against this for two reasons:
+
+- zero-initialization is not substantially more efficient than our API,
+  in particular because `memset` cannot be used (the OCaml
+  representation of 0 is actually a full-word 1, which is not periodic
+  at the byte level)
+  
+- in the future we could have immediate types for which the value 0 is
+  invalid, if we allow to "shift" the initial tag of a constructor.
+
+
+## Alternatives
+
+### Bigarrays
+
+One can use `(int, int_elt, c_layout) Bigarray.Array1.t` for this purpose.
+
+Downsides:
+
+- working with Bigarrays is a lot more code/ceremony (but this could be hidden
+  by a dedicated library)
+
+- bigarray access is slightly slower due to performing an extra read
+  to get the layout
```

Comment:
@lthls, @chambart pointed out that this is not true if the layout type of the bigarray is statically known. You get an extra indirection, but this should be okay. On the other hand, you might still run in GC-pacing issues if you have lots of small bigarrays.